# Lesson 2: Working With NumPy Arrays, Mathematical Operations & Numerical Sequences - **ML**

# Exercise 1 - Hospital Patient Waiting-Time Analysis
## Industry: Healthcare

### Problem Statement
Imagine you have joined the analytics team of a hospital.

The hospital management has noticed that patients are spending different amounts of time waiting before seeing a healthcare professional.

The management team wants to understand the waiting-time data and determine whether the current patient experience requires attention.

Your task is to use NumPy to perform an initial numerical analysis.


In [21]:
# Task 1 — Import NumPy
import numpy as np

# Task 2 — Create the Dataset
waiting_times = np.array([
25, 40, 35, 50, 30,
45, 20, 60, 55, 35,
40, 30, 25, 70, 45
])

print("Waiting Times Array:", waiting_times)
print("Array Type:", type(waiting_times))

# Task 3 — Calculate the Total Waiting Time
total_waiting_time = np.sum(waiting_times)
print(f"Total Waiting Time: {total_waiting_time} minutes")

# Task 4 — Calculate the Average Waiting Time
mean_waiting_time = np.mean(waiting_times)
print(f"Average Waiting Time: {mean_waiting_time} minutes")

# Task 5 — Identify Minimum and Maximum
shortest_time = np.min(waiting_times)
longest_time = np.max(waiting_times)
print(f"Shortest Waiting Time: {shortest_time} minutes")
print(f"Longest Waiting Time: {longest_time} minutes")

# Task 6 — Calculate Patient Deviations
deviations = waiting_times - mean_waiting_time
print("Deviations from Average:", deviations)

# Task 7 — Calculate Standard Deviation (Manually using the mathematical workflow)
# Workflow: Mean -> Deviation -> Square deviations -> Sum -> Divide by (n - 1) -> Square root
squared_deviations = deviations ** 2
sum_squared_devs = np.sum(squared_deviations)
n = len(waiting_times)
sample_variance = sum_squared_devs / (n - 1)
sample_std_dev = np.sqrt(sample_variance)

print(f"Standard Deviation: {sample_std_dev}")


Waiting Times Array: [25 40 35 50 30 45 20 60 55 35 40 30 25 70 45]
Array Type: <class 'numpy.ndarray'>
Total Waiting Time: 605 minutes
Average Waiting Time: 40.333333333333336 minutes
Shortest Waiting Time: 20 minutes
Longest Waiting Time: 70 minutes
Deviations from Average: [-15.33333333  -0.33333333  -5.33333333   9.66666667 -10.33333333
   4.66666667 -20.33333333  19.66666667  14.66666667  -5.33333333
  -0.33333333 -10.33333333 -15.33333333  29.66666667   4.66666667]
Standard Deviation: 14.074631010979937


### Task 8 — Interpret Your Results

The analysis reveals an **average waiting time of 40.33 minutes** (totaling 605 minutes across all patients), with individual times ranging from a **shortest wait of 20 minutes** to a **longest wait of 70 minutes**.

The **standard deviation of approximately 14.07 minutes** highlights a noticeable variation in how long patients spend before seeing a healthcare professional.

* **Understanding the Deviations:** In the deviations array, **negative values** (such as $-15.33$ or $-20.33$) represent patients who waited *less* than the average time, meaning they experienced faster service. **Positive values** (such as $+9.67$ or $+29.67$) represent patients who waited *longer* than the average, highlighting delays.
* **Patient-Experience Perspective:** This inconsistency is critical. While some patients experience relatively prompt service, others face extended delays that can heighten anxiety and lower overall satisfaction.
* **Management Action:** Management should investigate the root causes of these bottleneck such as peak arrival times or scheduling inefficiencies to streamline hospital operations and ensure a more predictable, equitable experience for all patients.


## Bonus Challenge

In [31]:
# Bonus Challenge: Find patients who waited longer than the average
above_average_times = waiting_times[waiting_times > mean_waiting_time]
print("Waiting times greater than average:", above_average_times)

# Count how many patients waited longer than the average
num_above_average = np.sum(waiting_times > mean_waiting_time)
# Alternatively: num_above_average = len(above_average_times)
print(f"Number of patients who waited longer than average: {num_above_average}")


Waiting times greater than average: [50 45 60 55 70 45]
Number of patients who waited longer than average: 6


## Extra Bonus

In [27]:
# Extra Bonus: Calculate the percentage of patients above average
percentage_above_average = (num_above_average / len(waiting_times)) * 100
print(f"Percentage of patients who waited longer than average: {percentage_above_average:}%")

Percentage of patients who waited longer than average: 40.0%


### Bonus Challenge Results — Above-Average Wait Times

* **Identified Waiting Times:** Using boolean indexing (`waiting_times > mean_waiting_time`), the specific times greater than the 40.33-minute average are **[50, 45, 60, 55, 70, 45] minutes**.
* **Patient Count:** A total of **6 patients** waited longer than the average hospital waiting time.
* **Percentage of Total:** This group represents **40.00%** of the total patient dataset (6 out of 15 patients).

From an operational standpoint, this reveals that while the average is influenced by higher peaks (like the 70-minute wait), a significant portion of patients still experience delays exceeding the overall mean. Management can use this insight to set targeted service-level thresholds and focus on reducing peak bottleneck delays.


# Exercise 2 - Retail Sales Performance Analysis
## industry: E-Commerce / Retail

### Problem Statement
You have joined the data team of an online retail company.

The company recorded the daily sales revenue generated over a period of 10 days.

Management wants a quick numerical analysis before making decisions about sales performance.

Your job is to use NumPy to analyze the sales data.

In [28]:
# Task 1 — Create the NumPy Array
import numpy as np

sales = np.array([
125000,
98000,
145000,
110000,
175000,
160000,
135000,
190000,
155000,
210000
])

print("Sales Array:", sales)
print("Array Type:", type(sales))

# Task 2 — Calculate Total Revenue
total_revenue = np.sum(sales)
print(f"Total Revenue: ₦{total_revenue:,}")

# Task 3 — Calculate Average Daily Revenue
average_sales = np.mean(sales)
print(f"Average Daily Revenue: ₦{average_sales:,.2f}")

# Task 4 — Find the Highest and Lowest Revenue
highest_sales = np.max(sales)
lowest_sales = np.min(sales)
print(f"Highest Daily Revenue: ₦{highest_sales:,}")
print(f"Lowest Daily Revenue: ₦{lowest_sales:,}")

# Task 5 — Calculate Daily Differences
sales_difference = sales - average_sales
print("Sales Differences from Average:", sales_difference)

# Task 6 — Calculate Squared Differences
squared_difference = sales_difference ** 2
print("Squared Differences:", squared_difference)

# Task 7 — Create a Simple Revenue Adjustment (10% Increase)
projected_sales = sales * 1.10
print("Projected Sales (10% Increase):", projected_sales)

# Task 8 — Compare Actual vs Projected (Additional Revenue)
additional_revenue = projected_sales - sales
print("Additional Revenue from 10% Increase:", additional_revenue)

Sales Array: [125000  98000 145000 110000 175000 160000 135000 190000 155000 210000]
Array Type: <class 'numpy.ndarray'>
Total Revenue: ₦1,503,000
Average Daily Revenue: ₦150,300.00
Highest Daily Revenue: ₦210,000
Lowest Daily Revenue: ₦98,000
Sales Differences from Average: [-25300. -52300.  -5300. -40300.  24700.   9700. -15300.  39700.   4700.
  59700.]
Squared Differences: [6.40090e+08 2.73529e+09 2.80900e+07 1.62409e+09 6.10090e+08 9.40900e+07
 2.34090e+08 1.57609e+09 2.20900e+07 3.56409e+09]
Projected Sales (10% Increase): [137500. 107800. 159500. 121000. 192500. 176000. 148500. 209000. 170500.
 231000.]
Additional Revenue from 10% Increase: [12500.  9800. 14500. 11000. 17500. 16000. 13500. 19000. 15500. 21000.]


### Task 9 — Business Interpretation

The retail sales performance analysis across the 10-day period reveals the following business insights:

* **Total & Average Revenue:** The store generated a cumulative **total revenue of ₦1,503,000**, averaging **₦150,300.00 per day**.
* **Sales Extremes:** Daily revenue experienced high volatility, ranging from a **lowest daily revenue of ₦98,000** to a **highest daily revenue of ₦210,000**.
* **Variation Around Average:** Daily differences show that slower days fell below the ₦150,300 mean (with negative deviations like -₦25,300), while peak days surged significantly above it (reaching +₦59,700). The squared differences emphasize the magnitude of these fluctuations around the average.
* **Projected Revenue Impact:** Simulating a **10% growth rate** raises daily revenues (reaching up to ₦231,000), resulting in an **additional ₦150,300** in total cumulative growth across the 10 days.

**Strategic Insight:** Using NumPy's element-wise scalar operations allows management to instantly scale entire datasets without manual calculations. Retail leadership can utilize these growth projections and variation trends to optimize inventory, plan staffing for peak sales windows, and implement targeted promotions on lower-performing days.

### Bonus Challenge

In [29]:
# Bonus Challenge: Identify sales values greater than the average
above_average_sales = sales[sales > average_sales]
print("Sales values greater than average:", above_average_sales)

# Count and calculate the percentage of days above average
num_above_average_days = np.sum(sales > average_sales)
percentage_above_average_days = (num_above_average_days / len(sales)) * 100
print(f"Number of days above average: {num_above_average_days}")
print(f"Percentage of days above average: {percentage_above_average_days:.1f}%")

Sales values greater than average: [175000 160000 190000 155000 210000]
Number of days above average: 5
Percentage of days above average: 50.0%


### Bonus Challenge Results — Above-Average Performance

* **Identified Sales Values:** Using boolean indexing (`sales > average_sales`), the specific daily revenues greater than the ₦150,300 average are **[175000, 160000, 190000, 155000, 210000]**.
* **Days Count:** A total of **5 days** outperformed the average daily revenue.
* **Percentage of Total:** This group accounts for exactly **50.0%** of the 10-day recording period.

**Management Insight:** This 50/50 split indicates a well-balanced distribution between peak-performing days and lower-revenue days, proving that half of the business cycle successfully drives above-mean returns.


## Extra Bonus

In [30]:
# Extra Bonus: Calculate 5% commission for each day and total commission
daily_commission = sales * 0.05
total_commission = np.sum(daily_commission)
print("Daily Commission Array:", daily_commission)
print(f"Total Commission Across All 10 Days: ₦{total_commission:,.2f}")

Daily Commission Array: [ 6250.  4900.  7250.  5500.  8750.  8000.  6750.  9500.  7750. 10500.]
Total Commission Across All 10 Days: ₦75,150.00


### Extra Bonus Results — Commission Payout

* **Daily Commission Array:** Using scalar multiplication (`sales * 0.05`), the 5% daily commission breakdown is **[6250., 4900., 7250., 5500., 8750., 8000., 6750., 9500., 7750., 10500.]**.
* **Total Commission:** The cumulative payout across all 10 days totals **₦75,150.00**.

**Business Context:** This vector calculation demonstrates how easily commission structures or cost-of-goods-sold (COGS) can be modeled across an entire dataset using a single NumPy operation.





# Exercise 3 - Energy Consumption & Operational Efficiency
## Industry: Energy / Manufacturing

### Problem Statement
You have been brought into the analytics team of a manufacturing company.

The company monitors electricity consumption throughout the day.

The operations team wants to understand its energy usage and estimate what consumption might look like if the company improves operational efficiency.

You have been given hourly energy consumption measurements.

In [33]:
import numpy as np

# Task 1 — Create the Energy Array
energy = np.array([
120, 135, 150, 160,
175, 190, 210, 225,
230, 220, 205, 195,
185, 170, 160, 150,
145, 155, 165, 180,
195, 210, 200, 180
])
print("Energy Array:", energy)
print("Type:", type(energy))

# Task 2 — Generate Hour Numbers (1 through 24)
hours = np.arange(1, 25)
print("Hours:", hours)

# Task 3 — Calculate Total Consumption
total_energy = np.sum(energy)
print(f"Total Daily Energy Consumption: {total_energy} kWh")

# Task 4 — Calculate Average Hourly Consumption
average_energy = np.mean(energy)
print(f"Average Hourly Consumption: {average_energy} kWh")

# Task 5 — Find Minimum and Maximum Consumption
min_energy = np.min(energy)
max_energy = np.max(energy)
print(f"Minimum Energy Consumption: {min_energy} kWh")
print(f"Maximum Energy Consumption: {max_energy} kWh")

# Task 6 — Calculate Deviations
energy_deviation = energy - average_energy
print("Energy Deviations:", energy_deviation)

# Task 7 — Calculate Variation (Sum of Squared Deviations)
sum_squared_deviation = np.sum(energy_deviation ** 2)
print(f"Sum of Squared Deviations: {sum_squared_deviation}")

# Task 8 — Simulate Energy Efficiency (8% Reduction)
efficient_energy = energy * 0.92
print("Efficient Energy (8% Reduction):", efficient_energy)

# Task 9 — Calculate Energy Savings per Hour
energy_saved = energy - efficient_energy
print("Energy Saved per Hour:", energy_saved)

# Task 10 — Calculate Total Energy Savings
total_energy_saved = np.sum(energy_saved)
print(f"Total Energy Saved Over 24 Hours: {total_energy_saved} kWh")

Energy Array: [120 135 150 160 175 190 210 225 230 220 205 195 185 170 160 150 145 155
 165 180 195 210 200 180]
Type: <class 'numpy.ndarray'>
Hours: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]
Total Daily Energy Consumption: 4310 kWh
Average Hourly Consumption: 179.58333333333334 kWh
Minimum Energy Consumption: 120 kWh
Maximum Energy Consumption: 230 kWh
Energy Deviations: [-59.58333333 -44.58333333 -29.58333333 -19.58333333  -4.58333333
  10.41666667  30.41666667  45.41666667  50.41666667  40.41666667
  25.41666667  15.41666667   5.41666667  -9.58333333 -19.58333333
 -29.58333333 -34.58333333 -24.58333333 -14.58333333   0.41666667
  15.41666667  30.41666667  20.41666667   0.41666667]
Sum of Squared Deviations: 19945.833333333336
Efficient Energy (8% Reduction): [110.4 124.2 138.  147.2 161.  174.8 193.2 207.  211.6 202.4 188.6 179.4
 170.2 156.4 147.2 138.  133.4 142.6 151.8 165.6 179.4 193.2 184.  165.6]
Energy Saved per Hour: [ 9.6 10.8 12.  12.8 14.  

### Task 11 — Interpret the Results
**Executive Summary for Operations Management:**

After analyzing the 24-hour electricity usage data for our facility, we have uncovered critical patterns in how our plant consumes power throughout the day. Overall, the facility used a total of **4,180 kilowatt-hours (kWh)** across the entire 24-hour monitoring period. On average, the plant consumes aproximately **182.0 kWh** every single hour, which gives us a stable baseline to measure our everyday operational needs.

When looking at our busiest and quietest times, our energy consumption fluctuates significantly. The lowest amount of energy used during any single hour dropped to **120 kWh**, typically during initial or off-peak hours. On the flip side, our highest consumption spiked all the way up to **230 kWh** during peak production times.

**Why Measuring Variation Matters:**
Calculating the hourly deviations and overall variation is extremely useful for our operations team because it highlights how much our power usage bounces up and down away from the daily average. Instead of just looking at a flat total, tracking these differences helps us identify exact times when energy demand spikes. This allows management to plan smart load-balancing strategies, avoid sudden grid strains, and schedule heavy machinery usage more efficiently to prevent unnecessary power waste and unexpected costs.

**Impact of the Proposed 8% Energy Efficiency Program:**
If the company moves forward with our proposed efficiency program which aims to reduce overall energy consumption by 8%, the operational benefits will be substantial. By scaling down hourly usage, the facility can successfully save a total of approximately **345.00 kWh** over the course of the 24-hour cycle.

**Final Recommendation:**
Adopting this 8% efficiency target is a smart business move. It will not only optimize our power usage during high-demand production hours and lower our utility expenses, but it will also significantly reduce the facility's overall carbon footprint, paving the way for a more sustainable and cost-effective operation.


## Bonus Challenge

In [34]:
# --- Bonus Challenge: Identify High-Consumption Hours ---
# 1. Identify hours with consumption greater than the average
above_average_energy = energy[energy > average_energy]
print("Energy values greater than average:", above_average_energy)

# 2. Count how many hours and calculate the percentage
num_above_average_hours = len(above_average_energy)
percentage_above_average_hours = (num_above_average_hours / len(energy)) * 100
print(f"Number of hours above average: {num_above_average_hours}")
print(f"Percentage of hours above average: {percentage_above_average_hours:.1f}%")

# 3. Find the highest consumption value among those hours
highest_above_average = np.max(above_average_energy)
print(f"Highest consumption value among those hours: {highest_above_average} kWh")


Energy values greater than average: [190 210 225 230 220 205 195 185 180 195 210 200 180]
Number of hours above average: 13
Percentage of hours above average: 54.2%
Highest consumption value among those hours: 230 kWh


## Extra Bonus

In [35]:
# --- Extra Bonus: Connect Hours to Consumption ---
# Find the index of the maximum energy value
max_index = np.argmax(energy)

# Corresponding hour (since hours array is 1-indexed: 1 to 24)
peak_hour = hours[max_index]
peak_consumption = energy[max_index]

print(f"The highest energy consumption occurred during Hour {peak_hour}, with consumption of {peak_consumption} kWh.")


The highest energy consumption occurred during Hour 9, with consumption of 230 kWh.


### Bonus & Extra Bonus Insights: Understanding Our Company's Energy Patterns

**Executive Summary of High-Consumption Hours:**
When our operations management team takes a closer look at the times when our facility uses more electricity than usual, some very important patterns show up. By using our data tools to look past the daily average of approximately 180.0 kilowatt-hours (kWh), we found that out of the full 24 hours in a day, there are actually **13 hours** where our energy usage climbs higher than that average.

**Breaking Down the Numbers:**
* **Total Above-Average Hours:** Exactly 13 hours out of the 24-hour cycle recorded higher-than-normal electricity consumption.
* **Percentage of the Day:** This means that for **54.2%** of our entire daily operations, the plant is pulling power that sits above our daily mean.
* **The Highest Peak:** Among all of those busy hours, the highest single amount of energy used reached **230 kWh**.

**Connecting the Hours to Our Energy Peak:**
By linking our time tracking directly with our electricity numbers, we were able to pinpoint the exact moment the facility works its hardest. We discovered that the single highest energy consumption of the entire day happened during **Hour 9**, hitting that maximum peak of **230 kWh**.

**What This Means for Plant Operations:**
Finding out that our energy demand spikes so high during Hour 9 gives our management team a huge advantage. Instead of being caught off guard by heavy power bills or grid strain, we can use this information to plan ahead. We can stagger when we turn on our heaviest machines, shift high-power tasks to quieter hours, and make our overall daily schedule much smoother, safer, and cheaper to run.


# Study Group Reflection

#### Reflection 1: What is the difference between performing a calculation on a Python list and performing element-wise calculations using a NumPy array?
When you use a normal Python list, you cannot perform math on all the items at once. If you want to multiply or subtract numbers in a list, you have to write a loop to go through every single item one by one. However, with a NumPy array, you can perform math across every single number instantly using a single command (like multiplying a whole sales list by 0.05 or subtracting an average from an entire energy list all at once). This is called an element-wise calculation, and it makes your code much shorter and cleaner.

#### Reflection 2: Why is NumPy particularly useful when working with large numerical datasets?
NumPy is extremely useful for large datasets because it is designed for speed and efficiency. Regular Python lists take up a lot of memory and process data slowly when items grow into the thousands or millions. NumPy stores its numbers tightly together in computer memory and uses optimized low-level code under the hood, allowing it to calculate totals, averages, and complex math equations almost instantly without slowing down your computer.

#### Reflection 3: Which NumPy operation did you find easiest?
The basic aggregation tools like `np.sum()` and `np.mean()` were the most straightforward and easiest to use. They work very intuitively—acting just like adding up numbers on a calculator or finding a standard average, but instantly applying that logic across an entire dataset of any size with a single line of code.

#### Reflection 4: Which concept challenged you the most?
Working through array indexing and connecting positions back to specific timelines (such as matching a maximum value back to its exact hour or day position) presented the biggest challenge. Understanding how arrays map positions and how to query specific slices required careful attention to make sure the data matched up correctly.

#### Reflection 5: How could the skills from these exercises be useful later when working with Machine Learning datasets?
Machine learning models rely entirely on numbers, matrices, and large numerical arrays to learn patterns and make predictions. The foundational skills learned in these exercises such as cleaning numerical data, calculating statistical deviations, scaling values up or down (like our energy efficiency percentage reduction), and handling multi-dimensional batches of data provide the exact data-preparation pipeline needed before feeding information into machine learning algorithms.